src/doc/trpl/closures.md

```diff
@@ -1,214 +1,510 @@
+(expected &-ptr,
+    found closure) [E0308]
+         |x| x + num
+         ^~~~~~~~~~~
+
+```
+
+This error is letting us know that we don't have a `&'static Fn(i32) -> i32`,
+we have a `[closure <anon>:7:9: 7:20]` Wait, what?
+
+Because each closure generates its own environment `struct` and implementation
+of `Fn` and friends, these types are anonymous. They exist just solely for
+this closure. So Rust shows them as `closure <anon>`, rather than some
+autogenerated name.
+
+But why doesn't our closure implement `&'static Fn`? Well, as we discussed before,
+closures borrow their environment. And in this case, our environment is based
+on a stack-allocated `5`, the `num` variable binding. So the borrow has a lifetime
+of the stack frame. So if we returned this closure, the function call would be
+over, the stack frame would go away, and our closure is capturing an environment
+of garbage memory!
+
+So what to do? This _almost_ works:
+
+```rust,ignore
+fn factory() -> Box<Fn(i32) -> i32> {
+    let num = 5;
+
+    Box::new(|x| x + num)
 }
+# fn main() {
+let f = factory();
+
+let answer = f(1);
+assert_eq!(6, answer);
+# }
+```
+
+We use a trait object, by `Box`ing up the `Fn`. There's just one last problem:
+
+```text
+error: `num` does not live long enough
+Box::new(|x| x + num)
+         ^~~~~~~~~~~
 ```
 
-You might ask yourself: why do we need to introduce two type
-parameters `F` and `G` here?  Evidently, both `f` and `g` have the
-same signature: `Fn(i32) -> i32`.
+We still have a reference to the parent stack frame. With one last fix, we can
+make this work:
 
-That is because in Rust each closure has its own unique type.
-So, not only do closures with different signatures have different types,
-but different closures with the *same* signature have *different*
-types, as well!
+```rust
+fn factory() -> Box<Fn(i32) -> i32> {
+    let num = 5;
 
-You can think of it this way: the behavior of a closure is part of its
-type.  Therefore, using a single type parameter for both closures
-will accept the first of them, rejecting the second. The distinct
-type of the second closure does not allow it to be represented by the
-same type parameter as that of the first.  We acknowledge this, and
-use two different type parameters `F` and `G`.
+    Box::new(move |x| x + num)
+}
+# fn main() {
+let f = factory();
 
-This also introduces the `where` clause, which lets us describe type
-parameters in a more flexible manner.
+let answer = f(1);
+assert_eq!(6, answer);
+# }
+```
 
-That's all you need to get the hang of closures! Closures are a little bit
-strange at first, but once you're used to them, you'll miss them
-in other languages. Passing functions to other functions is
-incredibly powerful, as you will see in the following chapter about iterators.
+By making the inner closure a `move Fn`, we create a new stack frame for our
+closure. By `Box`ing it up, we've given it a known size, and allowing it to
+escape our stack frame.
```

Comment:
Recommendation: "When should I take which kind of closure?"

My personal intuition on this:

* `FnOnce` is the most general to *accept*: both `Fn` and `FnMut` can be used as `FnOnce`.  If you're only going to call a closure once, use `FnOnce`.  The downside is that the caller *must* have *unique ownership* of the closure in question.  Also, `FnOnce` doesn't work \[easily] with dynamic dispatch.
* `FnMut` is the next most general: `Fn` can be used as `FnMut`, but `FnOnce` cannot.  Use `FnMut` if you have no particular reason *not* to.
* `Fn` is the least general, but strongly indicates to the caller that the closure should be "pure": that is, it shouldn't have side-effects (this can be worked around with interior mutability).  Use `Fn` if you have to (you are accessing the closure through an immutable borrow), or you want to discourage impure closures.